Add new command `jj config unset`

pylbrecht opened the pull request:
Allow unsetting config values similar to `git config unset`.

```bash
$ jj config set --user some-key some-val
$ jj config get some-key
some-val
$ jj config unset --user some-key
$ jj config get some-key
Config error: configuration property "some-key" not found
For help, see https://martinvonz.github.io/jj/latest/config/.
```

Unsetting a key, which is part of a table, might leave that table empty. For now we do not delete such empty tables, as there may be cases where an empty table is semantically meaningful (https://github.com/martinvonz/jj/issues/4458#issuecomment-2407109269).

For example:

```toml
[table]
key = "value"

[another-table]
key = "value"
```

Running `jj config unset --user table.key` will leave us with `table` being empty:
```toml
[table]

[another-table]
key = "value"
```

---

Closes #4458

# Checklist
If applicable:

- [x] I have updated CHANGELOG.md
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

pylbrecht commented on cli/src/commands/config/unset.rs:
```diff
@@ -0,0 +1,36 @@
+        )));
+    }
+
+    // TODO(pylbrecht): do we need to check_wc_author() here?
```
I saw we have this check here:
https://github.com/martinvonz/jj/blob/f4ec653e12ddaeb39c98124dad7d2de5845321d2/cli/src/commands/config/set.rs#L66-L72

Do we need to inform the user in the same way for `jj config unset`?

yuja replied:
Maybe no. It's technically correct that unsetting repo config may change the user name/email, but I expect such situation is rare.

pylbrecht commented on cli/src/commands/config/unset.rs:
```diff
@@ -0,0 +1,36 @@
+use crate::config::remove_config_value_from_file;
+use crate::config::ConfigNamePathBuf;
+
+/// Update config file to unset the given option.
```
Should we document for users that unsetting keys, which are part of a table, might result in empty tables?

yuja replied:
I don't think explicit doc is needed.

pylbrecht commented on cli/tests/test_config_command.rs:
```diff
@@ -602,6 +602,96 @@ fn test_config_set_nontable_parent() {
+    test_env.set_config_path(user_config_path.clone());
+    let repo_path = test_env.env_root().join("repo");
+
+    test_env.jj_cmd_ok(&repo_path, &["config", "set", "--user", "foo", "true"]);
+    test_env.jj_cmd_ok(&repo_path, &["config", "unset", "--user", "foo"]);
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "set", "--user", "table.foo", "true"],
+    );
+    test_env.jj_cmd_ok(&repo_path, &["config", "unset", "--user", "table.foo"]);
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "set", "--user", "inline-table", "{ foo = true }"],
+    );
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "unset", "--user", "inline-table.foo"],
+    );
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &[
+            "config",
+            "set",
+            "--user",
+            "table.inline-table",
+            "{ foo = true }",
+        ],
+    );
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "unset", "--user", "table.inline-table.foo"],
+    );
```
I am unsure how extensive we want to be in terms of coverage here. I am fine with adding more cases, if you feel it's beneficial.

yuja replied:
I think these tests are good enough. CLI tests are slow.

pylbrecht commented on cli/tests/test_config_command.rs:
```diff
@@ -602,6 +602,96 @@ fn test_config_set_nontable_parent() {
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "init", "repo"]);
+    let repo_path = test_env.env_root().join("repo");
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "set", "--repo", "test-key", "test-val"],
+    );
+    test_env.jj_cmd_ok(&repo_path, &["config", "unset", "--repo", "test-key"]);
```
I just covered the very simple happy path for `--repo`, since the implementation covers both `--user` and `--repo`. However, I don't mind copy/pasting the cases from `test_config_unset_for_user()` if that is beneficial.

(I am not sure how likely it is that the implementations for `--user` and `--repo` diverge in the future, such that having extensive tests for both becomes necessary.)

pylbrecht commented:
I would like to get some feedback on my changes. I left some open questions as inline comments.
Also, since writing Rust still feels pretty wonky, I am happy for any Rust specific feedback.

yuja commented on cli/src/commands/config/mod.rs:
```diff
@@ -17,8 +17,11 @@ mod get;
 
 use tracing::instrument;
+use unset::cmd_config_unset;
+use unset::ConfigUnsetArgs;
```
nit: `self::unset` for consistency (and clarity)

pylbrecht replied:
Thanks, my LSP took care of that auto-import and I wasn't aware that it didn't do the right thing here.

yuja commented on cli/src/config.rs:
```diff
@@ -583,6 +584,58 @@ fn read_config_path(config_path: &Path) -> Result<config::Config, config::Config
+            format!("Failed to parse file {path}", path = path.display()),
+            err,
+        )
+    })?;
```
nit: Perhaps, we can extract read/write helper functions or a function that takes callback `update_doc: impl FnOnce(&mut Document) -> Result<..>`.

pylbrecht replied:
I really liked the idea of an `update_doc` callback, but it turned out to be a little bumpy. `write_config_value_to_file()` takes a `value` argument, whereas `remove_config_value_from_file()` does not. I tried making `value` optional, but that made the interface awkward.  

That's when I decided to go for `read_config()` and `write_config()` helpers, despite not being too happy about it. 

Since I do not consider myself an advanced Rustacean, feel free to give it a shot yourself; I might be missing something.

pylbrecht replied:
I just realized I baked the refactoring and the feature into one change. I will split these into two separate ones.

yuja commented on cli/src/config.rs:
```diff
@@ -583,6 +584,58 @@ fn read_config_path(config_path: &Path) -> Result<config::Config, config::Config
+            .ok_or(config::ConfigError::NotFound(key.to_string()))
+            .and_then(|table| {
+                table
+                    .as_table_like_mut()
```
nit: better to use `.as_table_mut()` for consistency. Think inline table as a single value.

pylbrecht replied:
Good point, thanks!

Should we give people a hint in the error message if they try to unset a key of an inline table then?
Or is `"{key} is not a table"` good enough?

(Maybe I am overthinking this.)

pylbrecht replied:
Great, happy to leave the `"{key} is not a table"`. In case people get confused, we can special-case it then.

yuja commented on cli/tests/test_config_command.rs:
```diff
@@ -602,6 +602,96 @@ fn test_config_set_nontable_parent() {
+    );
+
+    let user_config_toml = std::fs::read_to_string(&user_config_path)
+        .unwrap_or_else(|_| panic!("Failed to read file {}", user_config_path.display()));
```
nit: just do `.unwrap()` in tests unless there's a special consideration

pylbrecht replied:
I just copy/pasted that from another test in the same file and haven't really thought about it.
Now that I put some thought into it, I guess the "custom" panic here serves a better error message than a plain `.unwrap()`, so I am inclined to leave it as is.

`.unwrap_or_else(...)`
```
---- test_config_command::test_config_unset_for_user stdout ----
thread 'test_config_command::test_config_unset_for_user' panicked at cli/tests/test_config_command.rs:665:29:
Failed to read file /private/var/folders/hb/zr1tp04d29b8wdh14lczb3c40000gn/T/jj-test-IkBUbD/config/config.toml
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

vs.

`.unwrap()`
```
---- test_config_command::test_config_unset_for_user stdout ----
thread 'test_config_command::test_config_unset_for_user' panicked at cli/tests/test_config_command.rs:664:64:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
---
Another option would be to use `.expect()`
```suggestion
        .expect(&format!("Failed to read file {}", user_config_path.display()));
```
I don't know if that is "better", because of `&format!(...)`. What's your take on this?

yuja replied:
Well, if `std::fs::read_to_string(&user_config_path)` is likely to fail, it makes sense to provide a better error message. However, we're testing "config unset" (which isn't the command to create new config file), so I would just use `.unwrap()`.

pylbrecht replied:
Thanks for elaborating. That makes a lot of sense now, since I copied that part from a `jj config set` test case. 💡 
I will replace it with `.unwrap()` then.

joyously replied:
Do you test unsetting it twice in a row? What would be the expected result for the second time?

pylbrecht replied:
> Do you test unsetting it twice in a row? What would be the expected result for the second time?

Right now there is no test case, which covers running `jj config unset` twice in a row for the same key.

With the current implementation, given this user config:
```toml
[table]
foo = "bar"
```
..running `jj config --user unset table.foo` twice in a row will result in the following error:
```
Config error: configuration property "table.foo" not found
For help, see https://martinvonz.github.io/jj/latest/config/.
```
..because the first time running `jj config --user unset table.foo` removes `table.foo`. So the second execution can't find the key any longer.

(The same applies for `jj config --repo unset table.foo`)

---

@joyously, your question got me thinking. I added a comment in the issue: https://github.com/martinvonz/jj/issues/4458#issuecomment-2438613512

yuja replied:
I don't know whether it should be an error or a warning, but it's nice that `jj config unset` reports unknown key to unset (i.e. noop change).

yuja reviewed (commented):
Looks generally good to me, thanks.

pylbrecht commented:
@yuja, thanks for your quick feedback!

I will address your comments over the next couple of days and let you know when I'm done.

pylbrecht commented:
@yuja, thanks again for taking the time to look through my changes. I addressed your comments, so the changes are ready for another review. For your sake, I hope this is the last round trip. 🤞 

I also added a test case for unsetting keys of inline tables, making it more explicit that we do not support this at the moment.
(When it comes to tests, I am biased towards "better too much than too little", so let me know if the overhead of another test is not worth it in this case.)

yuja commented on CHANGELOG.md:
```diff
@@ -78,6 +78,10 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+* New command `jj config unset` that unsets config values. For example,
+  `jj config unset --user user.name`.
+
+
```
nit: one blank line, not two

yuja commented on cli/src/commands/config/unset.rs:
nit: can you add "Copyright 2024 ..." comment to this file?

yuja commented on cli/src/config.rs:
```diff
@@ -633,13 +645,35 @@ pub fn write_config_value_to_file(
+            .ok_or_else(|| config::ConfigError::NotFound(key.to_string()))
+            .and_then(|table| {
+                table.as_table_mut().ok_or_else(|| {
+                    config::ConfigError::Message(format!("\"{key}\" is not a table",))
```
nit: remove trailing comma

(at your option) we can use `r#".."` to remove `\"` escapes.

yuja reviewed (approved):
LGTM, thanks.
